### DISCUSSION TITLE
Dynamic header on Chromium PDF

### DISCUSSION THREAD

**Discussion by tassin-gauthier** [gotenberg/gotenberg#416] [Q&A] [answered]
I would like to insert a header on every page of the pdf generated with chromium, with a dynamic content. That's mean the header change for each pdf, i cannot set a fixed height for all the pdf to generate.

I did a lot of research, with and without gotenberg, and this is what i have found : 

- use : 
```css
@page {
    @top-center {
        content: element(header)
    }

#header {
    position: running(header);
}
```
and try to apply margin into `@top-center`. It doesn't work because chrome doesn't support page-margin boxes. [Link](https://github.com/Fyrd/caniuse/issues/4034#issuecomment-404836879
). 

- use thead / tbody, which is the most promoted solution : 
```html
<table>
   <thead>header</thead>
   <tbody>content</tbody>
</table>
```
And this working well, the content is not overlapping, you don't need to add weird margin... until the header reach around 256px. After that the header is not repeated on the next pages. It only appear on the first page. 

Unfortunately, this is a chromium bug and it will not be fixed soon... [link](https://bugs.chromium.org/p/chromium/issues/detail?id=962435#c13)

- use : 

```css
#header {
   position: fixed;
   top: 0;
}
```

Unfortunately, it overlaps with the content and you cannot add margin anywhere. If you add margin to the `@page` into css or with gotenberg, it will add a margin for the whole page and the header with still overlap with the content. If you add margin to the content (not the header), it will work for the first page but the header and the content will still overlap on the other pages.

- Finally and this is the best solution for now : inject my header into a header.html file and send it as gotenberg documentation said [here](https://gotenberg.dev/docs/modules/chromium). But the content is overlapping with the header. To avoid that, The solution provided is to add a "marginTop" into the post request... which is work pretty good for a fixed header ! But as i said in the first paragraph : my height is dynamic, i can't set a fixed marginTop.

Any suggestion ? Can i get the same result with another html/css/js tricks i didn't tried ? Can gotenberg handle the dynamic marginTop case in a future release ? 

How gotenberg succeed to add a marginTop which push the content but not the header ? If i understand how, maybe i could implements it myself and use some javascript to make this margin dynamic with the header height. 

Any help is appreciated, this is a major feature for a tool which generate 100.000+ pdf / day, thanks !

**Comment by gulien** [comment, marked as answer]
(Moving this to the discussions space)

Hello @tassin-gauthier,

Actually, I use the Chrome API to add the margin(s), so there is no black magic from Gotenberg here 😄 

[Paged.js](https://pagedjs.org/) might be the best bet for your use case. See for instance the [header/footer documentation](https://pagedjs.org/documentation/7-generated-content-in-margin-boxes/#running-elements%3A-headers%2Ffooter-with-specific-(complex)-content).

**Reply by tassin-gauthier** [reply]
Hello @gulien 

I tried [Paged.js](https://pagedjs.org/) and ... with some adjustments, this i working very well ! Thank you so much for that 😄 !

There is just a little problem with this solution and gotenberg. This is a simple example : https://gist.github.com/tassin-gauthier/9aad38823ec2e44913123c7c60621502

If you open this into chrome **and check the print preview**, you will see that everything is working : the header with an example of 500px is displayed on the head of every pages and the red border fit perfectly with no margin. (in real world, the height value is not forced to 500px like that, it's just for the example).

Now if you send the exact same file to gotenberg with a marginTop set to 0, the pdf add an unwanted margin : 

<img width="1070" alt="Capture d’écran 2022-02-03 à 15 56 43" src="https://user-images.githubusercontent.com/49563367/152367651-f10786d1-fd5e-4b29-8acc-5a092a5398c3.png">

If I remove the 
```html
<script src="https://unpkg.com/pagedjs/dist/paged.polyfill.js"></script>
```
of course the header is not repeated on every pages on print preview, but the margin disappears.

Do you know how can i remove this unwanted top margin ? Thanks !

**Reply by gulien** [reply]
@tassin-gauthier now that's weird 🤔  what if you set all the margins to 0?

**Reply by tassin-gauthier** [reply]
Ok i figured out : all the margin passed to gotenberg into the request body are not taken into account if you include paged.js.

And this : 
```css
@page {
   size: A4;
}
```
combine to appropriate paperWidth and paperHeight into the request body (paperWidth = 8.27 and paperHeight = 11.69) fixed the issue ! No more unwanted margin 🚀.

Thanks a lot 😄

**Reply by gulien** [reply]
@tassin-gauthier Good to know 😄 I was confident that the margins would override any properties from the page, but it seems not.

Glad it worked out for you!
